// react-touch-loader/lib/react-touch-loader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('./touch-loader.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var STATS = {
    init: '',
    pulling: 'pulling',
    enough: 'pulling enough',
    refreshing: 'refreshing',
    refreshed: 'refreshed',
    reset: 'reset',

    loading: 'loading' };

// pull to refresh
// tap bottom to load more
// loading more
exports.default = _react2.default.createClass({
    displayName: 'react-touch-loader',

    getInitialState: function getInitialState() {
        return {
            loaderState: STATS.init,
            pullHeight: 0
        };
    },
    getDefaultProps: function getDefaultProps() {
        return {
            distanceToRefresh: 60
        };
    },
    setInitialTouch: function setInitialTouch(touch) {
        this._initialTouch = {
            clientY: touch.clientY
        };
    },
    calculateDistance: function calculateDistance(touch) {
        return touch.clientY - this._initialTouch.clientY;
    },
    // 拖拽的缓动公式 - easeOutSine
    easing: function easing(distance) {
        // t: current time, b: begInnIng value, c: change In value, d: duration
        var t = distance;
        var b = 0;
        var d = screen.availHeight; // 允许拖拽的最大距离
        var c = d / 2.5; // 提示标签最大有效拖拽距离

        return c * Math.sin(t / d * (Math.PI / 2)) + b;
    },
    canRefresh: function canRefresh() {
        return this.props.onRefresh && [STATS.refreshing, STATS.loading].indexOf(this.state.loaderState) < 0;
    },

    touchStart: function touchStart(e) {
        if (!this.canRefresh()) return;
        if (e.touches.length == 1) this._initialTouch = {
            clientY: e.touches[0].clientY,
            scrollTop: this.refs.panel.scrollTop
        };
    },
    touchMove: function touchMove(e) {
        if (!this.canRefresh()) return;
        var scrollTop = this.refs.panel.scrollTop;
        var distance = this.calculateDistance(e.touches[0]);

        if (distance > 0 && scrollTop <= 0) {
            var pullDistance = distance - this._initialTouch.scrollTop;
            var pullHeight = this.easing(pullDistance);
            if (pullHeight) e.preventDefault(); // 减弱滚动

            this.setState({
                loaderState: pullHeight > this.props.distanceToRefresh ? STATS.enough : STATS.pulling,
                pullHeight: pullHeight
            });
        }
    },
    touchEnd: function touchEnd() {
        if (!this.canRefresh()) return;
        var endState = {
            loaderState: STATS.reset,
            pullHeight: 0
        };

        if (this.state.loaderState == STATS.enough) {
            // refreshing
            this.setState({
                loaderState: STATS.refreshing,
                pullHeight: 0
            });

            // trigger refresh action
            this.props.onRefresh(function () {
                // resove
                this.setState({
                    loaderState: STATS.refreshed,
                    pullHeight: 0
                });
            }.bind(this), function () {
                // reject
                this.setState(endState); // reset
            }.bind(this));
        } else this.setState(endState); // reset
    },

    loadMore: function loadMore() {
        this.setState({ loaderState: STATS.loading });
        this.props.onLoadMore(function () {
            // resolve
            this.setState({ loaderState: STATS.init });
        }.bind(this));
    },
    render: function render() {
        var _props = this.props;
        var className = _props.className;
        var hasMore = _props.hasMore;
        var initializing = _props.initializing;
        var _state = this.state;
        var loadMoreState = _state.loadMoreState;
        var loaderState = _state.loaderState;
        var pullHeight = _state.pullHeight;


        var footer = hasMore ? _react2.default.createElement(
            'div',
            { className: 'tloader-footer' },
            _react2.default.createElement('p', { className: 'tloader-btn', onClick: this.loadMore }),
            _react2.default.createElement(
                'p',
                { className: 'tloader-loading' },
                _react2.default.createElement('i', { className: 'ui-loading' })
            )
        ) : null;

        var style = pullHeight ? {
            WebkitTransform: 'translate3d(0,' + pullHeight + 'px,0)'
        } : null;

        var progressClassName = '';
        if (initializing > 0) progressClassName += ' progress';
        if (initializing > 1) progressClassName += ' ed';

        return _react2.default.createElement(
            'div',
            { ref: 'panel', className: 'tloader state-' + loaderState + ' ' + className + progressClassName, onTouchStart: this.touchStart, onTouchMove: this.touchMove, onTouchEnd: this.touchEnd },
            _react2.default.createElement(
                'div',
                { className: 'tloader-symbol' },
                _react2.default.createElement(
                    'p',
                    { className: 'tloader-msg' },
                    _react2.default.createElement('i', null)
                ),
                _react2.default.createElement(
                    'p',
                    { className: 'tloader-loading' },
                    _react2.default.createElement('i', { className: 'ui-loading' })
                )
            ),
            _react2.default.createElement(
                'div',
                { className: 'tloader-body', style: style },
                this.props.children
            ),
            footer
        );
    }
});
